### PR TITLE
Set up pre-commit to provide automatic formatting and static code checking.

### PR DESCRIPTION
**Pull request by mikedlr**
Currently set up with just the defaults - this sets up pre-commit  so that we can get automatic checking of code as we commit it to the repo.

### CODE REVIEW

**Comment by eegeeZA** [comment]
I like this, it will keep the code tidy and reduce future PR diffs. But it does look like this will need to be applied across all the code (once-off) for the PR to succeed.

**Comment by mikedlr** [comment]
@eegeeZA yes - that's right.  I currently have a patch containing the simple automated fixes from `perltidy` and the other utilities in the default setup which touches 203 files.  When I also cover the fixes required for `perlcritic` to be happy that will probably touch almost all of the files in the repo. 

I think I will write a BBS post which will warn people that this is coming - it will be much easier for any outstanding code and fixes to be merged before this happens.  I'm at least waiting for @vasile-gh to finish his current work on the cloning scripts first.

**Comment by jgonggrijp** [comment]
Will this actually run as a pre-commit hook?

**Comment by jgonggrijp** [comment]
Ah, found the answer at pre-commit.com. It _will_ run as a pre-commit hook, but only after a developer has run `pre-commit install` once.

File: docs/dev-notes.md [hunk 1]
```diff
@@ -26,6 +27,28 @@ should get you a shell inside a docker container with all of
...
+#####################
+
+We use Git as our source code repository management system. This is quite
+powerful and sophisticated and so sometimes a little intimidating to start
+with. Despite that, what you need to contribute to spamgourmet will be quite
+simple. If you don't know git then this is likely a good place to start
+
+https://docs.github.com/en/get-started/using-git
+
+For any changes you make,
```
Comment (jgonggrijp) [inline docs/dev-notes.md:40]: is some text missing here?

File: docs/dev-notes.md [hunk 2]
```diff
@@ -80,4 +102,25 @@ perl script to create testing data
...
+
+If you are going to do more than a couple of lines of changes then
+it's very likely to be useful to set up the configuation locally to
+give you warnings before committing your code to git.
+
+1. get a python install on your system including pip
+2. install pre-commit - https://pre-commit.com/#install
+3. run
+
+pre-commit install
```
Comment (jgonggrijp) [inline docs/dev-notes.md:126]:
```suggestion
3. run `pre-commit install`
```

File: docs/formatting-and-static-code-checks.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+In order to make it maximally easy to work on spamgourmet we are using static
+code checking and opinionated. The primary tool of choice for this is
```
Comment (jgonggrijp) [inline docs/formatting-and-static-code-checks.md:2]:
```suggestion
code checking and opinionated formatting. The primary tool of choice for this is
```

File: docs/formatting-and-static-code-checks.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+automate builds system (currently GitHub actions)
+
+Line Width handling
+###################
+
+Our agreed policy on line widths is approximately:
+
+- aim for lines less than 80 characters wide in English text
+  (e.g. documentation)
+- aim for lines less than 100 characters wide in code
```
Comment (jgonggrijp) [inline docs/formatting-and-static-code-checks.md:14]: Can we just combine the first two points into "aim for lines less than 80 characters"? It seems soft enough that people can exceed it by their own judgment.

File: docs/perl-style-guide.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+Use of other libraries needs to be considered with some care to ensure
+maintainability. Many Perl modules which used to be great seem to have lost
+their maintainers.
+
+Using a library is normally better for maintenance than rolling our own code,
+however please look through our discussion about libraries and if you want to
+use a new one, please mention it there, preferably after making some
+investigation to see if the library is still accepting bug fixes.
+
+https://github.com/spamgourmet/spamgourmet/issues/44
```
Comment (jgonggrijp) [inline docs/perl-style-guide.md:29]:
```suggestion
Using a library is normally better for maintenance than rolling our own code,
however please look through our [discussion][issue-14] about libraries and if you want to
use a new one, please mention it there, preferably after making some
investigation to see if the library is still accepting bug fixes.

[issue-14]: https://github.com/spamgourmet/spamgourmet/issues/44
```

File: .pre-commit-config-build.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# this is the set of checks that will be run during build and will
```
Comment (fredericmora) [inline .pre-commit-config-build.yaml:1]: Please have mercy on my OCD and capitalize the word at the start of a sentence (except if it's a program name)!

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# this is the full set of static code checks that we have defined that
```
Comment (fredericmora) [inline .pre-commit-config.yaml:1]: `This is`

File: docs/dev-notes.md (hunk 1 above)
Comment (fredericmora) [reply docs/dev-notes.md:40]: Remove that line unless you want to expand on it?

File: docs/dev-notes.md [hunk 8]
```diff
@@ -80,4 +102,25 @@ perl script to create testing data
...
+We use opinionated formatting and some static code quality
+checks. These run in GitHub and in order for a patch to be accepted
+into spamgourmet we'll need to get them to pass.
+
+For making a tiny, obvious change it's probably easiest to make the
+change and submit a patch. Watch the feedback after the build process
+and fix any problems that are reported until the build passes.
+
+If you are going to do more than a couple of lines of changes then
+it's very likely to be useful to set up the configuation locally to
```
Comment (fredericmora) [inline docs/dev-notes.md:119]: I'd go a bit further: 

```
If you are going to do more than a couple of lines of changes then
please set up the configuration locally to give you warnings before committing your code to git.```

**Review by fredericmora — APPROVED** [review]
I left a few small comments.

**Comment by mikedlr** [comment]
This is merged (but done with a rebase and then --ff-onlyi so GitHub didn't automatically mark as such)

**Pull request closed without merging**